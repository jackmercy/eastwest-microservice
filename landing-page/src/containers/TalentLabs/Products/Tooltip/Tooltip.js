import React, { memo, useCallback, useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import useStyles from './tooltip.styles';

const Tooltip = ({
  visible,
  offsetX,
  offsetY,
  sideOffset,
  className,
  children,
}) => {
  const ref = useRef();
  const classes = useStyles();
  const [state, setState] = useState({ clientX: 0, clientY: 0 });

  const getMousePosition = useCallback(({ clientX, clientY }) => {
    setState({ clientX, clientY });
  }, []);

  const addListener = useCallback(() => {
    if (typeof window !== 'undefined') {
      window.addEventListener('mousemove', getMousePosition);
    }
  }, [getMousePosition]);

  const removeListener = useCallback(() => {
    if (typeof window !== 'undefined') {
      window.removeEventListener('mousemove', getMousePosition);
      setState({ clientX: 0, clientY: 0 });
    }
  }, [getMousePosition]);

  useEffect(() => removeListener, []);

  useEffect(() => {
    if (visible) {
      addListener();
    } else {
      removeListener();
    }
  }, [visible]);

  const getTopPosition = () => {
    if (typeof window !== 'undefined' && ref.current) {
      const { height } = ref.current.getBoundingClientRect();
      if (state.clientY > window.innerHeight - height) {
        return state.clientY - height - offsetY;
      }
      return state.clientY + offsetY;
    }

    return state.clientY + offsetY;
  };

  const getLeftPosition = () => {
    if (typeof window !== 'undefined' && ref.current) {
      const { width } = ref.current.getBoundingClientRect();
      if (state.clientX - width / 2 + offsetX - sideOffset < 0) {
        return sideOffset;
      }
      if (
        state.clientX + width / 2 + offsetX + sideOffset >
        window.innerWidth
      ) {
        return window.innerWidth - width - offsetX - sideOffset;
      }

      return state.clientX - width / 2 + offsetX;
    }

    return state.clientX + offsetX;
  };

  return (
    <div
      ref={ref}
      className={classNames(classes.root, className)}
      style={{
        visibility:
          visible &&
          state.clientX !== 0 &&
          state.clientY !== 0 &&
          ref.current !== undefined
            ? 'visible'
            : 'hidden',
        top: getTopPosition(),
        left: getLeftPosition(),
      }}
    >
      {children}
    </div>
  );
};
Tooltip.propTypes = {
  children: PropTypes.node.isRequired,
  visible: PropTypes.bool,
  offsetX: PropTypes.number,
  offsetY: PropTypes.number,
  // Offset when element got left/right edge
  sideOffset: PropTypes.number,
  className: PropTypes.string,
};
Tooltip.defaultProps = {
  visible: true,
  offsetX: 0,
  offsetY: 0,
  sideOffset: 16,
  className: '',
};

export default memo(Tooltip);
