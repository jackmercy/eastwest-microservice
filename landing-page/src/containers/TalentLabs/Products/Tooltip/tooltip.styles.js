import { createUseStyles } from 'react-jss';

export default createUseStyles({
  root: { display: 'block', position: 'fixed', zIndex: 99 },
});
