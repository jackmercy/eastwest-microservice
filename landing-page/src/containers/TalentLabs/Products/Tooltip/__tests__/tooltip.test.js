import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Tooltip from '../Tooltip';

describe('Testing TalentLabs > Products > Tooltip', () => {
  const props = {
    children: <span>Some content</span>,
  };

  it('should render match snapshot', () => {
    const wrapper = shallow(<Tooltip {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
