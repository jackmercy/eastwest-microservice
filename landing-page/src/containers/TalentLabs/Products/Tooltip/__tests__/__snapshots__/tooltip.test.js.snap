// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TalentLabs > Products > Tooltip should render match snapshot 1`] = `
<div
  className="root-0-2-1"
  style={
    Object {
      "left": 0,
      "top": 0,
      "visibility": "hidden",
    }
  }
>
  <span>
    Some content
  </span>
</div>
`;
